src/main/kotlin/com/ampnet/auditornode/persistence/repository/impl/LocalIpfsRepository.kt

```diff
@@ -41,6 +53,49 @@ class LocalIpfsRepository @Inject constructor(
     override fun fetchBinaryFileFromDirectory(directoryHash: IpfsHash, fileName: String): Try<IpfsBinaryFile> =
         fetchFileFromDirectory(directoryHash, fileName, ByteArray::class.java, ::IpfsBinaryFile)
 
+    override fun uploadFilesToDirectory(files: Flux<NamedIpfsFile>): Mono<Try<IpfsDirectoryUploadResponse>> {
+        return files.collectList().flatMap {
+            val requestBody = MultipartBody.builder()
+
+            it.forEach { upload -> requestBody.addPart("file", upload.fileName, upload.content) }
+
+            val request = HttpRequest.POST(
+                "http://localhost:${ipfsProperties.localClientPort}/api/v0/add" +
+                    "?quieter=true&wrap-with-directory=true&pin=true",
+                requestBody.build()
+            ).apply {
+                contentType(MediaType.MULTIPART_FORM_DATA_TYPE)
+            }
+
+            Mono.from(httpClient.retrieve(request, ByteArray::class.java))
+                .map<Try<IpfsDirectoryUploadResponse>> { response ->
+                    val responses = String(response).split("\n")
+                        .map(String::trim)
+                        .filter(String::isNotEmpty)
+                        .map(objectMapper::readTree)
+                        .mapNotNull { json ->
+                            val fileName = json["Name"]?.asText()
+                            val ipfsHash = json["Hash"]?.asText()?.let(::IpfsHash)
+
+                            if (fileName != null && ipfsHash != null) {
+                                IpfsFileUploadResponse(fileName, ipfsHash)
+                            } else {
+                                null
+                            }
+                        }
+
+                    val (directoryResponse, fileResponses) = responses.partition { r -> r.fileName.isEmpty() }
+
+                    if (directoryResponse.size != 1) {
+                        MissingUploadedIpfsDirectoryHash.left()
+                    } else {
+                        IpfsDirectoryUploadResponse(fileResponses, directoryResponse[0].ipfsHash).right()
+                    }
+                }
+                .onErrorResume { e -> Mono.just(IpfsHttpError(e).left()) }
+        }
+    }
```

Comment:
This is a bit long. Could you simplify somehow? Why did you switch to reactive?

Reply:
I will break it into smaller methods. I have no choice but to use reactive for this implementation because Micronaut only offers multi-file upload as a reactive stream.